cmd/lndinit/cmd_migrate_db.go

```diff
@@ -0,0 +1,595 @@
+			}
+
+			if err := copyBucketKvdb(srcBucket, destBucket); err != nil {
+				return fmt.Errorf("error copying bucket "+
+					"'%s': %v", loggableKeyName(k), err)
+			}
+
+			return nil
+		}
+
+		err := dest.Put(k, v)
+		if err != nil {
+			return fmt.Errorf("error copying key '%s': %v",
+				loggableKeyName(k), err)
+		}
+
+		return nil
+	})
+}
+
+func (x *migrateDBCommand) migrateEtcd(srcTx walletdb.ReadWriteTx,
+	prefix string) error {
+
+	ctx := context.Background()
+	cfg := x.Dest.Etcd.CloneWithSubNamespace(prefix)
+	destDb, ctx, cancel, err := etcd.NewEtcdClient(ctx, *cfg)
+	if err != nil {
+		return err
+	}
+	defer cancel()
+
+	err = srcTx.ForEachBucket(func(key []byte) error {
+		log("Copying top-level bucket '%s'", loggableKeyName(key))
+
+		return migrateBucketEtcd(
+			ctx, destDb, []string{string(key)},
+			srcTx.ReadWriteBucket(key),
+		)
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func openDb(cfg *DB, prefix string) (walletdb.DB, error) {
+	backend := cfg.Backend
+
+	var args []interface{}
+
+	graphDir := filepath.Join(cfg.Bolt.DataDir, "graph", cfg.Bolt.Network)
+	walletDir := filepath.Join(
+		cfg.Bolt.DataDir, "chain", "bitcoin", cfg.Bolt.Network,
+	)
+	towerServerDir := filepath.Join(
+		cfg.Bolt.TowerDir, "bitcoin", cfg.Bolt.Network,
+	)
+
+	switch backend {
+	case lncfg.BoltBackend:
+		var path string
+		switch prefix {
+		case lncfg.NSChannelDB:
+			path = filepath.Join(graphDir, lncfg.ChannelDBName)
+
+		case lncfg.NSMacaroonDB:
+			path = filepath.Join(walletDir, lncfg.MacaroonDBName)
+
+		case lncfg.NSDecayedLogDB:
+			path = filepath.Join(graphDir, lncfg.DecayedLogDbName)
+
+		case lncfg.NSTowerClientDB:
+			path = filepath.Join(graphDir, lncfg.TowerClientDBName)
+
+		case lncfg.NSTowerServerDB:
+			path = filepath.Join(
+				towerServerDir, lncfg.TowerServerDBName,
+			)
+
+		case lncfg.NSWalletDB:
+			path = filepath.Join(walletDir, lncfg.WalletDBName)
+		}
+
+		const (
+			noFreelistSync = true
+			timeout        = time.Minute
+		)
+
+		args = []interface{}{
+			path, noFreelistSync, timeout,
+		}
+		backend = kvdb.BoltBackendName
+		log("Opening bbolt backend at %s for prefix '%s'", path, prefix)
+
+	case kvdb.EtcdBackendName:
+		args = []interface{}{
+			context.Background(),
+			cfg.Etcd.CloneWithSubNamespace(prefix),
+		}
+		log("Opening etcd backend at %s with namespace '%s'",
+			cfg.Etcd.Host, prefix)
+
+	case kvdb.PostgresBackendName:
+		args = []interface{}{
+			context.Background(),
+			&postgres.Config{
+				Dsn: cfg.Postgres.Dsn,
+			},
+			prefix,
+		}
+		log("Opening postgres backend at %s with prefix '%s'",
+			cfg.Postgres.Dsn, prefix)
+
+	default:
+		return nil, fmt.Errorf("unknown backend: %v", backend)
+	}
+
+	return kvdb.Open(backend, args...)
+}
+
+func putKeyValueEtcd(ctx context.Context, cli *clientv3.Client, key,
+	value string) error {
+
+	ctx, cancel := context.WithTimeout(ctx, etcdTimeout)
+	defer cancel()
+
+	_, err := cli.Put(ctx, key, value)
```

Comment:
One quick low hanging speedup is to gather key/values to a buffer then when the buffer hits say 100-1000 items then flush it to etcd in one txn. It's doable since the individual keys don't depend on each other.